source/event_stream_rpc_server.c

```diff
@@ -0,0 +1,1107 @@
+    int error_code,
+    struct aws_channel *channel,
+    void *user_data) {
+    (void)bootstrap;
+
+    struct aws_event_stream_rpc_server_listener *server = user_data;
+
+    if (!error_code) {
+        AWS_LOGF_INFO(
+            AWS_LS_EVENT_STREAM_RPC_SERVER,
+            "id=%p: incoming connection with channel %p.",
+            (void *)server,
+            (void *)channel);
+        AWS_FATAL_ASSERT(channel && "Channel should never be null with a 0 error code.");
+
+        struct aws_event_stream_rpc_server_connection *connection = s_create_connection_on_channel(server, channel);
+
+        if (!connection) {
+            int error = aws_last_error();
+            server->on_new_connection(NULL, error, NULL, server->user_data);
+            aws_channel_shutdown(channel, error);
+        }
+
+        struct aws_event_stream_rpc_connection_options connection_options;
+        AWS_ZERO_STRUCT(connection_options);
+
+        aws_event_stream_rpc_server_connection_acquire(connection);
+        AWS_LOGF_TRACE(
+            AWS_LS_EVENT_STREAM_RPC_SERVER,
+            "id=%p: invoking on_new_connection with connection %p.",
+            (void *)server,
+            (void *)connection);
+
+        if (server->on_new_connection(connection, AWS_ERROR_SUCCESS, &connection_options, server->user_data)) {
+            aws_channel_shutdown(channel, aws_last_error());
+            aws_event_stream_rpc_server_connection_release(connection);
+            return;
+        }
+
+        AWS_FATAL_ASSERT(
+            connection_options.on_connection_protocol_message && "on_connection_protocol_message must be specified!");
+        AWS_FATAL_ASSERT(connection_options.on_incoming_stream && "on_connection_protocol_message must be specified");
+        connection->on_incoming_stream = connection_options.on_incoming_stream;
+        connection->on_connection_protocol_message = connection_options.on_connection_protocol_message;
+        connection->user_data = connection_options.user_data;
+        connection->bootstrap_owned = true;
+        aws_event_stream_rpc_server_connection_release(connection);
+
+    } else {
+        AWS_LOGF_ERROR(
+            AWS_LS_EVENT_STREAM_RPC_SERVER,
+            "id=%p: invoking on_new_connection with error %s",
+            (void *)server,
+            aws_error_debug_str(error_code));
+        server->on_new_connection(NULL, error_code, NULL, server->user_data);
+    }
+}
+
+/* this is just to get the connection object off of the channel. */
+static inline struct aws_event_stream_rpc_server_connection *s_rpc_connection_from_channel(
+    struct aws_channel *channel) {
+    struct aws_channel_slot *our_slot = NULL;
+    struct aws_channel_slot *current_slot = aws_channel_get_first_slot(channel);
+    AWS_FATAL_ASSERT(
+        current_slot &&
+        "It should be logically impossible to have a channel in this callback that doesn't have a slot in it");
+    while (current_slot->adj_right) {
```

Comment:
What if someone installs a protocol built on top of event stream RPC?

Reply:
we'll deal with it then. the fatal assert here is a great way for us to catch and adapt when it happens.